File: day1/day1.pl
:- module(day1, []).

:- use_module(library(pure_input)).
:- use_module(library(dcg/basics)).

input([X|Data]) -->
    integer(X),
    "\n",
    input(Data).

input([]) --> eos.

load_data(Data) :-
    open('day1/input.dat', read, Stream),
    phrase_from_stream(input(Data), Stream).

star(1, X) :-
    load_data(Numbers),
    member(A, Numbers),
    member(B, Numbers),
    A + B =:= 2020,
    X is A * B.

star(2, X) :-
    load_data(Numbers),
    member(A, Numbers),
    member(B, Numbers),
    member(C, Numbers),
    A + B + C =:= 2020,
    X is A * B * C.

:- begin_tests(day1).

test(star1) :- star(1, 988771), !.
test(star2) :- star(2, 171933104), !.

:- end_tests(day1).

File: day1/Day1.hs
module Day1 where

import Data.Maybe (mapMaybe)
import Text.Read ( readMaybe )

main :: IO ()
main = do
    file <- readFile "input.dat"
    let numbers = mapMaybe readMaybe (lines file)
    let pairs = cartProd numbers numbers
    let result = star pairs
    printResult result
    let tris = triProd numbers
    let result2 = star2 tris
    printResult result2


printResult :: Maybe Int -> IO ()
printResult Nothing = putStrLn "No hay solución"
printResult (Just x) = putStrLn ("Solución: " ++ show x)

cartProd :: [Int] -> [Int] -> [(Int, Int)]
cartProd xs ys = [(x,y) | x <- xs, y <- ys]

star :: [(Int, Int)] -> Maybe Int
star [] = Nothing
star ((x, y):xs) = if x + y == 2020 then Just (x * y) else star xs

triProd :: [Int] -> [(Int, Int, Int)]
triProd xs = [(x, y, z) | x <- xs, y <- xs, z <- xs]

star2 :: [(Int, Int, Int)] -> Maybe Int
star2 [] = Nothing
star2 ((x, y, z):xs) = if x + y + z == 2020 then Just (x * y * z) else star2 xs